src/librustc_mir/transform/const_prop.rs

```diff
@@ -755,14 +755,26 @@ struct CanConstProp {
     can_const_prop: IndexVec<Local, ConstPropMode>,
     // false at the beginning, once set, there are not allowed to be any more assignments
     found_assignment: IndexVec<Local, bool>,
+    // Cache of locals' information
+    local_kinds: FxHashMap<Local, LocalKind>,
```

Comment:
This can be an `IndexVec<Local, LocalKind>` to avoid hashing.

Reply:
Wait, aren't they read-only? I made a HashMap to allow for addition of Locals one by one during the loop that examines them.

How do `IndexVec<T, U>` work by the way? They seem really interesting.

Reply:
`IndexVec<T, U>` is basically just a normal `Vec<U>` except that instead of using `usize` to index into it, you use `T` where `T` is a "newtype wrapper" around an integer. It should be pretty straightforward to replace this `FxHashMap` with an `IndexVec`.

Reply:
Ohh, okay, I'll dig around to see if I can build it piece by piece :3

Reply:
You can take a look at how `can_const_prop` and `found_assignment` work since they are also `IndexVec<Local, _>`.

Reply:
I'll start with this today and then go on to work on the Block-based propagation :3

Reply:
I think I did it :)